sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp

```diff
@@ -448,3 +455,1201 @@ ZER_APIEXPORT zer_result_t ZER_APICALL zerPlatformGetInfo(
+                 ///< of devices.
+) {
+
+  PI_ASSERT(Platform, ZER_RESULT_INVALID_PLATFORM);
+
+  auto Res = Platform->populateDeviceCacheIfNeeded();
+  if (Res != ZER_RESULT_SUCCESS) {
+    return Res;
+  }
+
+  // Filter available devices based on input DeviceType.
+  std::vector<zer_device_handle_t> MatchedDevices;
+  std::shared_lock<pi_shared_mutex> Lock(Platform->PiDevicesCacheMutex);
+  for (auto &D : Platform->PiDevicesCache) {
+    // Only ever return root-devices from piDevicesGet, but the
+    // devices cache also keeps sub-devices.
+    if (D->isSubDevice())
+      continue;
+
+    bool Matched = false;
+    switch (DeviceType) {
+    case ZER_DEVICE_TYPE_ALL:
+      Matched = true;
+      break;
+    case ZER_DEVICE_TYPE_GPU:
+    case ZER_DEVICE_TYPE_DEFAULT:
+      Matched = (D->ZeDeviceProperties->type == ZE_DEVICE_TYPE_GPU);
+      break;
+    case ZER_DEVICE_TYPE_CPU:
+      Matched = (D->ZeDeviceProperties->type == ZE_DEVICE_TYPE_CPU);
+      break;
+    case ZER_DEVICE_TYPE_FPGA:
+      Matched = D->ZeDeviceProperties->type == ZE_DEVICE_TYPE_MCA;
+      break;
+    case ZER_DEVICE_TYPE_MCA:
+      Matched = D->ZeDeviceProperties->type == ZE_DEVICE_TYPE_MCA;
+      break;
+    default:
+      Matched = false;
+      zePrint("Unknown device type");
+      break;
+    }
+    if (Matched)
+      MatchedDevices.push_back(D.get());
+  }
+
+  uint32_t ZeDeviceCount = MatchedDevices.size();
+
+  if (*NumDevices == 0)
+    *NumDevices = ZeDeviceCount;
+  else {
+    auto N = std::min(ZeDeviceCount, *NumDevices);
+    std::copy_n(MatchedDevices.begin(), N, Devices);
+  }
+  return ZER_RESULT_SUCCESS;
+}
+
+ZER_APIEXPORT zer_result_t ZER_APICALL zerDeviceGetInfo(
+    zer_device_handle_t Device,  ///< [in] handle of the device instance
+    zer_device_info_t ParamName, ///< [in] type of the info to retrieve
+    size_t *pSize, ///< [in,out] pointer to the number of bytes needed to return
+                   ///< info queried. The call shall update it with the real
+                   ///< number of bytes needed to return the info
+    void *ParamValue ///< [out][optional] array of bytes holding the info.
+                     ///< If *pSize input is not 0 and not equal to the real
+                     ///< number of bytes needed to return the info then the
+                     ///< ::ZER_RESULT_ERROR_INVALID_SIZE error is returned and
+                     ///< pDeviceInfo is not used.
+) {
+  PI_ASSERT(Device, ZER_RESULT_INVALID_DEVICE);
+  UrReturnHelper ReturnValue(pSize, ParamValue);
+
+  ze_device_handle_t ZeDevice = Device->ZeDevice;
+
+  switch ((int)ParamName) {
+  case ZER_DEVICE_INFO_TYPE: {
+    switch (Device->ZeDeviceProperties->type) {
+    case ZE_DEVICE_TYPE_GPU:
+      return ReturnValue(ZER_DEVICE_TYPE_GPU);
+    case ZE_DEVICE_TYPE_CPU:
+      return ReturnValue(ZER_DEVICE_TYPE_CPU);
+    case ZE_DEVICE_TYPE_FPGA:
+      return ReturnValue(ZER_DEVICE_TYPE_FPGA);
+    default:

```

Comment:
should we also have MCA if we have it in zerDeviceGet?

Reply:
perhaps, no one would still be using it, so let's figure out later on